File: java/Solution39.java
package leetCode.java;

import java.util.*;

//Backtracking Knapsack O(n^(target / min(candidates))) O(target), n = candidates.length
class Solution39 {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtracking(0, new ArrayList<>(), candidates, target);
        return res;
    }

    public void backtracking(int start, List<Integer> selected, int[] nums, int target) {
        if (target == 0) {
            res.add(new ArrayList<>(selected));
            return;
        }
        if (target < 0) {
            return;
        }

        for (int i = start; i < nums.length; ++i) {
            selected.add(nums[i]);
            backtracking(i, selected, nums, target - nums[i]); // 從i繼續試探, 不是i+1
            selected.remove(selected.size() - 1);
        }
    }
}


class Solution39_2 {
    public List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backtracking(0, new ArrayList<Integer>(), candidates, target);
        return res;
    }

    public void backtracking(int i, List<Integer> list, int[] candidates, int target){
        if(target == 0){//先判斷成功避免i==candidates.length()時return
            res.add(new ArrayList<Integer>(list));//O(n)
            return;
        }
        if(i == candidates.length || target < 0) return;

        backtracking(i + 1, list, candidates, target);

        list.add(candidates[i]);
        backtracking(i, list, candidates, target - candidates[i]);
        list.remove(list.size() - 1);
    }
}
/* 背包問題, 可以重複 candidates = [2,3,6,7], target = 7
				[]
		2				[]
	22		2		3		[]
  222  22		  33  3    6  []
	...
*/

File: c++/Solution39.cpp
//Backtracking Knapsack O(n * 2^n) O(target), n = candidates.size()
class Solution39 {
public:
    vector<vector<int>> res;

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> selected;
        backtracking(0, selected, candidates, target);
        return res;
    }

    void backtracking(int start, vector<int>& selected, vector<int>& nums, int target) {
        if (target == 0) {
            res.push_back(selected);
            return;
        }
        if (target < 0) return;

        for (int i = start; i < nums.size(); ++i) { // 元素不相同, 所以不用判斷前一個是否相同
            selected.push_back(nums[i]);
            backtracking(i, selected, nums, target - nums[i]); // 從i繼續試探
            selected.pop_back();
        }
    }
};
/* candidates = [2,3,5], target = 8
_ -> 2 -> 22 -> 222 -> 2222(o)
                    -> 2223(x)
                    -> 2225(x)
             -> 223 -> 2233(x)
                    -> 2235(x)
             -> 225 -> 2255(x)
          23 -> 233(o)
             -> 235(x)
          25 -> 255(x)
     3 -> 33 -> 333(x)
             -> 335(x)
       -> 35(o)
     5 -> 55(x)

res=[[2,2,2,2],[2,3,3],[3,5]]
*/